src/p2p/Join/validate.ts

```diff
@@ -0,0 +1,306 @@
+}
+/**
+ * Makes sure that the given `joinRequestVersion` is not older than the
+ * current version of the node. If it is, it returns a `JoinRequestResponse`
+ * object with `success` set to `false` and `fatal` set to `true`. The `reason`
+ * field will contain a message describing the validation error.
+ *
+ * If the `joinRequestVersion` is not older than the current version of the
+ * node, it returns `null`.
+ */
+export function validateVersion(joinRequestVersion: string): JoinRequestResponse | null {
+  if (config.p2p.checkVersion && !isEqualOrNewerVersion(version, joinRequestVersion)) {
+    /* prettier-ignore */ warn(`version number is old. Our node version is ${version}. Join request node version is ${joinRequestVersion}`)
+    nestedCountersInstance.countEvent('p2p', `join-reject-version ${joinRequestVersion}`)
+    return BAD_VERSION_ERROR
+  } else return null
+}
+
+export const BAD_SIGNATURE_ERROR = {
+  success: false,
+  reason: 'Bad signature, sign owner and node attempted joining mismatched',
+  fatal: true,
+}
+/**
+ * Makes sure that the given `joinRequest` is signed by the node that is
+ * attempting to join. If it is not, it returns a `JoinRequestResponse` object
+ * with `success` set to `false` and `fatal` set to `true`. The `reason` field
+ * will contain a message describing the validation error.
+ *
+ * If the `joinRequest` is signed by the node that is attempting to join, it
+ * returns `null`.
+ */
+export function verifyJoinRequestSigner(joinRequest: P2P.JoinTypes.JoinRequest): JoinRequestResponse | null {
+  //If the node that signed the request is not the same as the node that is joining
+  if (joinRequest.sign.owner != joinRequest.nodeInfo.publicKey) {
+    /* prettier-ignore */ warn(`join-reject owner != publicKey ${{ sign: joinRequest.sign.owner, info: joinRequest.nodeInfo.publicKey }}`)
+    nestedCountersInstance.countEvent('p2p', `join-reject owner != publicKey`)
+    return BAD_SIGNATURE_ERROR
+  }
+
+  return null
+}
+
+export const ALREADY_SEEN_ERROR = {
+  success: false,
+  reason: 'Node has already been seen this cycle. Unable to add join request.',
+  fatal: false,
+}
+/**
+ * Makes sure that the given `joinRequest`'s node  has not already been seen this
+ * cycle. If it has, it returns a `JoinRequestResponse` object with `success`
+ * set to `false` and `fatal` set to `false`. The `reason` field will contain a
+ * message describing the validation error.
+ *
+ * If the `joinRequest`'s node has not already been seen this cycle, it returns
+ * `null`.
+ */
+function verifyUnseen(publicKey: hexstring): JoinRequestResponse | null {
+  // Check if this node has already been seen this cycle
+  if (getSeen().has(publicKey)) {
+    if (logFlags.p2pNonFatal) nestedCountersInstance.countEvent('p2p', `join-skip-seen-pubkey`)
+    if (logFlags.p2pNonFatal) info(ALREADY_SEEN_ERROR.reason)
+    return ALREADY_SEEN_ERROR
+  }
+
+  // Mark node as seen for this cycle
+  getSeen().add(publicKey)
+
+  return null
+}
```

Comment:
The AI review mentioned this one, but it seems wrong. I think the single line ifs confuse it.  
> The function verifyUnseen in validate.ts adds public keys to the seen set even if they are already in it. This might lead to redundant operations and could be optimized.